rudder-license/src/main/scala/com/normation/license/License.scala

```diff
@@ -0,0 +1,724 @@
+*/
+
+package com.normation.license
+
+import org.joda.time.DateTime
+import java.nio.file.Files
+import java.nio.charset.Charset
+import scala.util.control.NonFatal
+import java.nio.file.Paths
+import java.security.Signature
+import java.security.interfaces.RSAPrivateKey
+import java.security.interfaces.RSAPublicKey
+import org.bouncycastle.openssl.PEMParser
+import org.bouncycastle.asn1.x509.SubjectPublicKeyInfo
+import java.security.spec.X509EncodedKeySpec
+import org.bouncycastle.asn1.pkcs.PKCSObjectIdentifiers
+import java.security.KeyFactory
+import org.joda.time.format.ISODateTimeFormat
+import ca.mrvisser.sealerate.values
+import java.security.Security
+import org.bouncycastle.jce.provider.BouncyCastleProvider
+import org.bouncycastle.openssl.PEMKeyPair
+import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter
+import org.bouncycastle.util.encoders.Hex
+import org.bouncycastle.asn1.pkcs.PrivateKeyInfo
+import java.security.KeyPair
+import java.security.spec.PKCS8EncodedKeySpec
+
+
+/*
+ * This file define services responsible to manage licenses in Rudder:
+ * - check if a file is correctly signed
+ * - create a signed license file
+ *
+ * It is not responsible for stoping something to run, or periodically check
+ * the validity of a license. Only its consistency.
+ */
+
+/*
+ * This is an example of a signed license file. There two parts:
+ * - the one between "----" (excluded) is the signed part grouping all
+ *   the license information.
+ * - all the remaining part are meta-data. For now, all the meta-data
+ *   are about signing: when signing happen, with which key, what is
+ *   actually the signature.
+ *
+ * - min version is included
+ * - max version is ALSO included. So one should issue licenses for
+ *   a range like [3.0.0, 3.99.99], *NOT* [3.0.0, 4.0.0]
+ *
+ * The content exemple is below (with " * " at the begining of lines to
+ * be removed and digest on multiline which is not in real files!
+ *
+ * header=rudder-license-v1
+ * algorithm=SHA512withRSA
+ * digest=5b2b184e560a6edb452931f391b54260f0e70517655502670d6600a0b0c42b227abc3
+ * ef688544ae8f831b22213a7a3e184cae5a408c41375805428af3c21b543d017ccae55fa1b235
+ * a2f750efbb629f58d01afd61572f98771cd11c90c6a9855ff500a8b4e6b56219f5d1c5febc46
+ * 4613c43980e9f9f4fa4276754ef82b4d78cefcf599ab413cd562138f920071f4e3f81f132594
+ * dc454ac37db97e95c6df491db225116c51d9378e43003847d9cdc1a9d2c19df20e748f0d89d8
+ * 5a5d8a53e5e26909594c8528f22c8fe053782485b2996cc352c37ab4c1a18e6491c5105f2714
+ * 6b09c61966a91dd421320c6d93580f502dca7a3c8a71b29b96985f611c2cbed
+ * digestdate=2017-06-02T16:10:53+02:00
+ * keyid=640F9347
+ * ---- signed information about license
+ * licensee=François ARMAND, NORMATION SAS
+ * softwareid=rudder-plugin-datasources
+ * minversion=1.0
+ * maxversion=99.0
+ * startdate=2017-06-01T00:00:00Z
+ * enddate=2018-05-31T00:00:00Z
+ * ----
+ */
+
+
+/**
+ * A structure that allows to keep a field key name to
+ * allow parsing/reading them (serialization is such a
```

Comment:
👍

Reply:
Just so that you cry: this is the quantity of NO boilerplate brought by circe+doobie (ok, in a toy project, but still, it's amazing! Not ONE line of parsing): https://github.com/fanf/scala-pet-store/blob/master/src/main/scala/io/github/pauljamescleary/petstore/endpoint/PetEndpoints.scala
(the whole project is interesting for learning)

Reply:
OK, the person must learn how to align things, but well...

Reply:
I could make a Scala project to learn people to align things, and they can learn me some category theory / functionnal magic